### PR TITLE
BDT Analysis Note V2

### PR DESCRIPTION
**Pull request by bregnery**
This PR includes all changes necessary for limits with the bdt-based search, all plotting scripts used during AN V2, and all shell scripts with commands to run the tests seen in AN V2.

Why WIP?
- The shell scripts are quite specific to my environment and made with documentation in mind rather than usability. This should be changed before merging.
- I think some of my plotting scripts are also unnecessary and could be merged together to be used in a more versatile way.
- The plotting scripts are also quite environment specific at the moment and this should definitely be improved.

Why open a PR now?
- Documentation, I need to more easily see what needs to be improved and added.

### CODE REVIEW

File: run_bias_study.sh [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+
+# Array of mMed values
+mMed_values=(200 250 300 350 400 450 500 550)
+
+# Outer loop: Generate the datacards (skip if only running fits)
+if [ "$run_only_fits" = false ]; then
+  for mMed in "${mMed_values[@]}"
+  do
+    # Generate datacards for the current mMed value with variable mDark and datacards_date
+    python3 cli_boosted.py gen_datacards --bkg hists/merged_${datacards_date}/bkg_sel-bdt\=0.65.json --sig hists/smooth_${datacards_date}/SVJ_s-channel_mMed-${mMed}_mDark-${mDark_value}_rinv-0p3_alpha-peak_MADPT300_13TeV-madgraphMLM-pythia8_sel-bdt\=0.65_smooth.json -p main
```
Comment (kpedro88) [inline run_bias_study.sh:36]: i don't think `-p main` actually does anything here anymore. datacards should automatically be generated with all known/enabled pdfs.

File: run_limit_asimov_sig_inj.sh [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+# Default values
+datacards_date="20240919"  # Manually defined date for generating datacards
+dc_date=$(date +%Y%m%d)    # Dynamically set today's date for the gentoys command
+scan_date=$(date +%Y%m%d)  # Dynamically set today's date for the fittoys command
+mDark_value="10"
+#run_only_fits=false        # Default to running all loops
+run_only_fits=false        # Default to running all loops
+
+# Parse command-line arguments
+while getopts "d:f" opt; do
```
Comment (kpedro88) [inline run_limit_asimov_sig_inj.sh:14]: add `sel` argument

File: plot_bias_study.py [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+import matplotlib.pyplot as plt
+import numpy as np
+
+ROOT.gROOT.SetBatch(True)  # Prevents ROOT from opening canvases
+
+def main():
+    mz = [200, 250, 300, 350, 400, 450, 500, 550]
+    
+    # Directories for r_inj = 0 and r_inj = 1
+    idir_rinj0 = '/uscms/home/bregnery/work/CMSSW_11_3_4/src/boosted/svj_limits/bias_test/siginj0/fitDiagnosticsObserveddc_SVJ_s-channel_mMed-'
```
[the remark this answers is not in the document]
Comment (bregnery) [reply plot_bias_study.py]: Added, needs testing

File: plot_bias_study.py [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+        
+        mean_rinj1.append(fit_rinj1.Parameter(1))
+        emean_rinj1.append(fit_rinj1.ParError(1))
+        sigma_rinj1.append(fit_rinj1.Parameter(2))
+        esigma_rinj1.append(fit_rinj1.ParError(2))
+
+        # Save the histograms as PDFs and PNGs
+        canvas = ROOT.TCanvas("canvas", "canvas", 800, 600)
+        histo_rinj0.Draw()
+        canvas.SaveAs(f"bias_siginj_rinj0_MZ_{mz[i]}_GeV.pdf")
```
[the remark this answers is not in the document]
Comment (bregnery) [reply plot_bias_study.py]: Added, needs testing

File: plot_self_test.py [hunk 5]
```diff
@@ -0,0 +1,88 @@
+import ROOT
```
[the remark this answers is not in the document]
Comment (bregnery) [reply plot_self_test.py:1]: Added, needs testing.

File: plot_bias_study.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (bregnery) [reply plot_bias_study.py]: Tested

File: plot_self_test.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (bregnery) [reply plot_self_test.py:1]: tested

File: plot_bias_study.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (bregnery) [reply plot_bias_study.py]: Tested.

File: run_bias_study.sh [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+
+# Default values
+datacards_date="20240919"  # Manually defined date for generating datacards
+dc_date=$(date +%Y%m%d)    # Dynamically set today's date for the gentoys command
+toys_date=$(date +%Y%m%d)  # Dynamically set today's date for the fittoys command
+mDark_value="10"
+run_only_fits=false        # Default to running all loops
+
+# Parse command-line arguments
+while getopts "d:f" opt; do
```
[the remark this answers is not in the document]
Comment (bregnery) [reply run_bias_study.sh:13]: Done

File: run_bias_study.sh (hunk 1 above)
Comment (bregnery) [reply run_bias_study.sh:36]: Done and testing.

File: run_limit_asimov_sig_inj.sh (hunk 2 above)
Comment (bregnery) [reply run_limit_asimov_sig_inj.sh:42]: Done and tested

**Comment by bregnery** [comment]
Need to update documentation in the readme and then I can submit the PR.

All code has tested and seems to work well

**Comment by bregnery** [comment]
@kpedro88 Now I feel happy with the code :) Let me know if there is more to change.

File: run_limit_asimov_sig_inj.sh [hunk 7]
```diff
@@ -0,0 +1,103 @@
...
+        --siginj) siginj="$2"; shift ;;
+        --mDark) mDark_value="$2"; shift ;;
+        --rinv) rinv_value="$2"; shift ;; 
+        --mMed_values) IFS=' ' read -r -a mMed_values <<< "$2"; shift ;;  # Parse mMed_values as array
+        *) echo "Unknown parameter passed: $1"; exit 1 ;;
+    esac
+    shift
+done
+
+# Parse command-line arguments
```
Comment (kpedro88) [inline run_limit_asimov_sig_inj.sh:41]: isn't this now redundant with the previous lines?

Comment (bregnery) [reply run_limit_asimov_sig_inj.sh:41]: Thanks, dumb mistake, I just forgot to delete it. Changing and testing now.

Comment (bregnery) [reply run_limit_asimov_sig_inj.sh]: Done

**Pull request merged**